### PR TITLE
[WIP] Usart implementation

### PR DESCRIPTION
**Pull request by david-sawatzke**
TODO:
- Improve Documentation
- Investigate swm & maybe force the user to enable it on 845

### CODE REVIEW

**Comment by hannobraun** [comment]
Oh man, I've been doing pretty much the same :-)

I need to get this done in time for RustFest (need it for the workshop). I might use or ignore your stuff, whatever turns out more convenient. Once I'm back, I'll take a closer look and properly merge both of our work.

**Comment by david-sawatzke** [comment]
@hannobraun The example works now, there were 2 issues:
- The pinout is misleading, 25 is the tx pin
- For some reason, the clock for swm needs to be enabled, even though it should be enable from the start

File: src/usart.rs [hunk 1]
```diff
@@ -447,6 +447,27 @@ impl Peripheral for pac::USART2 {
...
+    const INTERRUPT: Interrupt = Interrupt::PIN_INT6_USART3;
+
+    type Rx = swm::U3_RXD;
+    type Tx = swm::U3_TXD;
+}
+
+#[cfg(feature = "845")]
+impl Peripheral for pac::USART4 {
+    /// USART4 and PIN_INT7 share an interrupt, this may cause difficulties
+    /// when trying to use both at the same time
```
Comment (hannobraun) [inline src/usart.rs]: Same of above, might be worth putting that into the documentation of the respective field in `Peripherals`.

File: examples/usart.rs [hunk 2]
```diff
@@ -1,25 +1,26 @@
...
-
 #[entry]
 fn main() -> ! {
     let p = Peripherals::take().unwrap();
 
-    let mut swm    = p.SWM.split();
+    let mut swm = p.SWM.split();
     let mut syscon = p.SYSCON.split();
 
+    // TODO
```
Comment (hannobraun) [inline examples/usart.rs]:
```suggestion
    // For some reason, the clock for swm needs to be enabled, even though it
    // should be enable from the start.
```

(this is the text from your comment)

File: examples/usart.rs [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
+    swm.handle = swm
+        .handle
+        .disable(&mut syscon.handle)
+        .enable(&mut syscon.handle);
```
Comment (hannobraun) [inline examples/usart.rs]: I'm fine with merging it like this, given the comment is updated (suggestion above) and we open a separate issue to track this problem.

File: examples/usart.rs [hunk 4]
```diff
@@ -44,25 +51,71 @@ fn main() -> ! {
...
+    #[cfg(feature = "82x")]
+    let rx_pin = swm.pins.pio0_18.into_swm_pin();
+
+    #[cfg(feature = "845")]
+    // Make PIO0_25 and PIO0_24 available to the switch matrix API, by changing
+    // their state using `into_swm_pin`. This is required, because we're going
+    // to use the switch matrix to assigne the USART0 functions to those pins.
+    //
+    // WARNING: The pinout for the lpc845brk uses tx/rx as seen from the
+    // perspective from the serial adapter, so this is used the opposite way
```
Comment (hannobraun) [inline examples/usart.rs]: Oh man, that's one of these things... probably took ages to figure out :-)

**Comment by hannobraun** [comment]
@david-sawatzke By the way, I'm currently doing some LPC845-related work for a client, which means I have the time to take this over, if you want. I don't want to step on your toes, I'm just offering to get this over the finish line, in case you're busy.

File: examples/usart.rs (hunk 4 above)
Comment (david-sawatzke) [reply examples/usart.rs]: Yeah, together with the swm issue, this took quite a while. I'm really thankful the c example code did the right thing

**Comment by david-sawatzke** [comment]
@hannobraun Thanks for the review!

I'm sorry about the code formatting stuff, I probably didn't pay attention when doing that and kept auto formatting on.

The TODOs were replaced by (bare bones) documentation, which should be enough for now.

I'm not happy with the clockconfig stuff, so I'd be happy if you had a go at it (everything else is also fair game).

**Comment by hannobraun** [comment]
Thank you for the additional changes! Merging.

> I'm sorry about the code formatting stuff, I probably didn't pay attention when doing that and kept auto formatting on.

No problem. It's just not ideal, which is why I wanted to mention it.

> I'm not happy with the clockconfig stuff, so I'd be happy if you had a go at it (everything else is also fair game).

I'll take a look. I have some more immediate priorities right now, but I'll take a stab at it later.

**Pull request merged**